hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/WriteStatus.java

```diff
@@ -52,9 +55,9 @@ public class WriteStatus implements Serializable {
 
   private final HashMap<HoodieKey, Throwable> errors = new HashMap<>();
 
-  private final List<HoodieRecord> writtenRecords = new ArrayList<>();
+  private final List<HoodieRecordDelegate> writtenRecordDelegates = new ArrayList<>();
 
-  private final List<HoodieRecord> failedRecords = new ArrayList<>();
+  private final List<Pair<HoodieRecordDelegate, Throwable>> failedRecords = new ArrayList<>();
```

Comment:
we might need the original record for failed list. since we might want to log them. only for success, we don't need entire record. 
btw, looks like we do deflate the content on success records. So, we can also consider leaving it as is. It was an oversight from my side.

Reply:
we do check if write status has error, then only trace-log those errors which only contain key info and exception. failed records are never used within hudi. we should be ok leaving as is.

Reply:
Yeah, for success records tracking from HoodieRecords, keep the record as is should be the most efficient way, here we unify the objects to `HoodieRecordDelegate` is because there are some code engine specific code path that does not hold any HoodieRecords, still, I think the API is not that elegant and we should do some refactoring when can have solution to unify the HoodieRecords abstractions first.